packages/data/addon/services/navi-metadata.ts

```diff
@@ -0,0 +1,142 @@
+ */
+import Service, { inject as service } from '@ember/service';
+import EmberArray from '@ember/array';
+import { setOwner, getOwner } from '@ember/application';
+import { assert } from '@ember/debug';
+import { getDataSource, getDefaultDataSource, DataSource } from 'navi-data/utils/adapter';
+import NaviMetadataSerializer, { EverythingMetadataPayload } from 'navi-data/serializers/metadata/interface';
+import KegService, { KegRecord } from './keg';
+import MetadataModelRegistry from 'navi-data/models/metadata/registry';
+import NaviMetadataAdapter from '../adapters/metadata/interface';
+
+type RequestOptions = {
+  dataSourceName?: string;
+};
+
+export default class NaviMetadataService extends Service {
+  @service('keg') private keg!: KegService;
+
+  /**
+   * @property {Array} loadedDataSources - list of data sources in which meta data has already been loaded
+   */
+  loadedDataSources: Set<string> = new Set();
+
+  /**
+   * @param dataSourceType
+   * @returns  adapter instance for type
+   */
+  private adapterFor(dataSourceType: string): NaviMetadataAdapter {
+    return getOwner(this).lookup(`adapter:metadata/${dataSourceType}`);
+  }
+
+  /**
+   * @param dataSourceType
+   * @returns serializer instance for type
+   */
+  private serializerFor(dataSourceType: string): NaviMetadataSerializer {
+    return getOwner(this).lookup(`serializer:metadata/${dataSourceType}`);
+  }
+
+  private dataSourceFor(dataSourceName?: string): DataSource {
+    return dataSourceName ? getDataSource(dataSourceName) : getDefaultDataSource();
+  }
+
+  /**
+   * @param payload
+   * @param dataSourceName
+   */
+  private loadEverythingMetadataIntoKeg(payload: EverythingMetadataPayload, dataSourceName: string) {
+    this.loadMetadataForType('table', payload.tables, dataSourceName);
+    this.loadMetadataForType('dimension', payload.dimensions, dataSourceName);
+    this.loadMetadataForType('timeDimension', payload.timeDimensions, dataSourceName);
+    this.loadMetadataForType('metricFunction', payload.metricFunctions || [], dataSourceName);
+    this.loadMetadataForType('metric', payload.metrics, dataSourceName);
+    this.loadedDataSources.add(dataSourceName);
+  }
+
+  /**
+   * Loads metadata based on type
+   * @param type - type of metadata, table, dimension, or metric
+   * @param metadataObjects - array of metadata objects
+   */
+  private loadMetadataForType<K extends keyof MetadataModelRegistry>(
+    type: K,
+    metadataObjects: Array<KegRecord>,
+    dataSourceName: string
+  ): EmberArray<MetadataModelRegistry[K]> {
+    const owner = getOwner(this);
+    metadataObjects.forEach(obj => setOwner(obj, owner));
+    return this.keg.pushMany(`metadata/${type}`, metadataObjects, {
+      namespace: dataSourceName
+    }) as EmberArray<MetadataModelRegistry[K]>;
+  }
+
+  async loadMetadata(options: RequestOptions = {}) {
+    const { type: dataSourceType, name: dataSourceName } = this.dataSourceFor(options.dataSourceName);
```

Comment:
Should we require and assert `options.dataSourceName` exists? I think going forward it would be nice to always specify it

Then you could also just directly call `getDataSource`

Reply:
I was thinking when I replace bard-metadata service with the navi one, I would make `service.loadMetadata()` load all data sources? Thoughts?

Reply:
👍 I think I'm fine with that. Although we maybe want to think about how this would work for `all()` when no dataSourceName is given since we would have to search every namespace if we expected it to have the same behavior